app/src/main/java/com/github/dedis/student20_pop/OrganizerActivity.java

```diff
@@ -118,10 +125,39 @@ public void OnEventTypeSelectedListener(EventType eventType) {
     @Override
     public void onAddWitnessListener() {
         if (ActivityCompat.checkSelfPermission(this, Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED) {
-            showFragment(new ConnectFragment(R.id.fragment_container_organizer), ConnectFragment.TAG);
+            showFragment(new QRCodeScanningFragment(ADD_WITNESS), QRCodeScanningFragment.TAG);
         } else {
-            showFragment(new CameraPermissionFragment(R.id.fragment_container_organizer), CameraPermissionFragment.TAG);
+            showFragment(new CameraPermissionFragment(ADD_WITNESS), CameraPermissionFragment.TAG);
         }
         // TODO : Get witness id from the QR code, add witness to witness list and send info to backend
     }
+
+    @Override
+    public void onCameraNotAllowedListener(QRCodeScanningType qrCodeScanningType) {
+        showFragment(new CameraPermissionFragment(qrCodeScanningType), CameraPermissionFragment.TAG);
+    }
+
+    @Override
+    public void onQRCodeDetected(String url, QRCodeScanningType qrCodeScanningType) {
+        Log.i(TAG, "Received qrcode url : " + url);
+        switch (qrCodeScanningType) {
+            case ADD_ROLL_CALL:
+                //TODO
+                break;
+            case ADD_WITNESS:
+                //TODO
+                Log.d("DEBUG", "URL Received : " + url);
+                break;
+            case CONNECT_LAO:
+                showFragment(ConnectingFragment.newInstance(url), ConnectingFragment.TAG);
+                break;
+            default:
+                break;
+        }
+    }
+
+    @Override
+    public void onCameraAllowedListener(QRCodeScanningType qrCodeScanningType) {
+        showFragment(new QRCodeScanningFragment(qrCodeScanningType), QRCodeScanningFragment.TAG);
+    }
```

Comment:
Very similar code in MainActivity and OrganizerActivity, wouldn't it be possible to have these methods only in one place?

Reply:
Yes I totally agree. For onQRCodeDetected, as I understand it, the connect fragment should never be attached to OrganizerActivity but only to MainActivity. And the add Witness functionality is only reachable when you are an organizer so inversely it should be only in OrganizerActivity and not in MainActivity. But for the future, I thought maybe we want to do some different UI when an organizer tries to connect to an event, or I don't know. 

For camera interfaces, I could avoid that by creating abstract classes instead of interfaces, but I think it is better to keep the code managing the fragments inside the Activity.

Reply:
Ok, it's all good for now as we just prioritize on making it work